container/container_unix.go

```diff
@@ -253,29 +248,43 @@ func (container *Container) IpcMounts() []Mount {
 	return mounts
 }
 
-// SecretMount returns the mount for the secret path
-func (container *Container) SecretMount() *Mount {
-	if len(container.SecretReferences) > 0 {
-		return &Mount{
-			Source:      container.SecretMountPath(),
-			Destination: containerSecretMountPath,
-			Writable:    false,
+// SecretMounts returns the mounts for the secret path.
+func (container *Container) SecretMounts() []Mount {
+	var mounts []Mount
+	for _, r := range container.SecretReferences {
+		if r.File == nil {
+			continue
 		}
+		mounts = append(mounts, Mount{
+			Source:      container.SecretFilePath(*r),
+			Destination: getSecretTargetPath(r),
+			Writable:    false,
+		})
 	}
 
-	return nil
+	return mounts
 }
 
 // UnmountSecrets unmounts the local tmpfs for secrets
 func (container *Container) UnmountSecrets() error {
-	if _, err := os.Stat(container.SecretMountPath()); err != nil {
-		if os.IsNotExist(err) {
-			return nil
+	for _, r := range container.SecretReferences {
+		if r.File == nil {
+			continue
+		}
+		srcPath := container.SecretFilePath(*r)
+		if _, err := os.Stat(srcPath); err != nil {
+			if os.IsNotExist(err) {
+				return nil
+			}
+			return err
+		}
+
+		if err := detachMounted(srcPath); err != nil {
+			return err
 		}
-		return err
 	}
```

Comment:
@cpuguy83: I believe I've fixed this. I'm not seeing any errors or leaked mounts when I delete a container that uses secrets. I'd appreciate it if you could confirm on your side.